### ISSUE TITLE
Enable incremental search and highlight while typing a search term in command mode?

### ISSUE THREAD

**Issue by bangedorrunt** [vscode-neovim/vscode-neovim#329]
hi,
in neovim, we can do that with
```
set incsearch hl
set inccommand=split
```
but this doesn't work on *vscode-neovim*, i also installed [ haya14busa/is.vim](https://github.com/haya14busa/is.vim) but it couldn't help.

**Comment by asvetliakov** [comment]
`incsearch` works just fine. Check if you've messed something with extension HL configuration

```
`set inccommand=split`
```
can't work because it creates another window in neovim. This can't be mapped to vscode

**Comment by asvetliakov** [comment]
Need to force `set inccommand=nosplit` in extension options

**Comment by babygau** [comment]
> Need to force `set inccommand=nosplit` in extension options

it doesn't help, i tried 

`%s/search/replace` 

but increment highlight is not shown. however doing `/search` works for me.

**Comment by asvetliakov** [comment]
So few things are happening here:

1. Search mode doesn't work in command mode because it tries to insert `<tab>` for completion. (simple cases like `/abc` or `?abc` are excluded for completion). That i'll fix soon
2. I tried to enable `inccommand=nosplit` and everything is worked, even neovim sends `buf_lines_event` with removed text after typing `/` before `replace`, BUT it doesn't revert the text back. So sounds like upstream issue,  `inccommand=nosplit` can't be supported now.

**Comment by asvetliakov** [comment]
Actually for `nosplit` we can ignore buf change event and just use it for HL

**Comment by babygau** [comment]
thank you for your explain! hope that will be sorted 👍

**Comment by awilkins** [comment]
Incremental search with `/` works well for me, but `%s/` is really jumpy - it seems to re-search for the first hit every time you type a char.

**Comment by theol0403** [comment]
Closed in favor of #607

**Comment by theol0403** [comment]
Currently, it seems like the new extmark support shows the replaced text, but does not remove the original. It's possible improved extmark support is needed to make this work.

**Comment by theol0403** [comment]
Fixes #607 - the ability to not force config option should solve it.

**Comment by ollien** [comment]
> Incremental search with `/` works well for me, but `%s/` is really jumpy - it seems to re-search for the first hit every time you type a char.

This is interesting. In plain nvim, if I scroll somewhat down a file, then type `:%s/_` (or some other char that appears both before and after my cursor), my editor does not scroll to the first match (the window stays in place), but it does in vscode. Interestingly, neovim does send us an explicit `grid_cursor_goto` that points at the first match (in this case, line 11), as well as a `win_viewport` which does the same. 

![image](https://github.com/vscode-neovim/vscode-neovim/assets/977151/458742c9-ebf7-4e9f-a75d-9b8ba44104fb)

Perhaps it is an incorrect assumption to scroll to wherever neovim places the cursor? It feels almost wrong to say, but it's the only thing that makes sense to me...

**Comment by theol0403** [comment]
@ollien it's possible

Scrolling is the other most troublesome feature of this project. It is implemented on the VSCode side and doesn't sync very well to Neovim. This breaks many things, like C-u/C-d, and adds many hacks/workarounds.

Looking around the project and PRs, you can see some attempts over the last few years to improve scrolling, but nothing has made it over the finish line.

So yeah, scrolling is another whole can of worms and likely out of your scope for the highlight manager. 

TLDR we aren't properly listening to when nvim *actually* tells us to scroll, and scroll when the cursor moves. 

https://github.com/vscode-neovim/vscode-neovim/blob/f7f89382571dedb7e84aaadf0ccf917371066ce2/src/cursor_manager.ts#L415

**Comment by ollien** [comment]
I'm increasingly convinced this is a bug in nvim, but I can't prove it. Take this file as an example

https://gist.githubusercontent.com/nikcub/3833406/raw/97891e880b4ed6bd29d514b983669f611a2523a7/index.php

1. Put your cursor on line 87
2. Open a command pallette and type `%/p` (do not hit enter)
3. Notice your window scrolls to the top (where `p` is located the first time)

Doing this in both nvim-vscode and nvim, only nvim-vscode jumps to the top. However, the events seem quite clear from the events (see below) that nvim want us to navigate to the top of the file. My initial theory is that a subsequent `win_viewport` might be sent, and we might be handling the initial `win_viewport` event before a `flush` is sent, but I'm not seeing evidence of that.

![2024-05-16_13-45](https://github.com/vscode-neovim/vscode-neovim/assets/977151/80aaace6-3675-43ba-ac84-f69e6adf3352)

@justinmk I would love your input here. I would make an issue on the nvim issue tracker, but I don't have a great way to setup a MWE that excludes vscode-nvim from the equation (if you do have a way to produce one, I will!)

**Comment by theol0403** [comment]
> It seems every time a character is typed, we fully resend the input chars to nvim. This would explain the jumpiness as more chars are typed, but not when typing a single character.

Yeah, the command line is one of the least polished parts of the plugin but we are really limited by vscode here. 

Not really a good way of forwarding keyboard events, or responding to arrow keys, backspace, or any other editing of the cmdline. 

For that reason, the full text is sent for every update. But, that messes up suggestions and such.

**Comment by xiyaowong** [comment]
Wasn't this issue fixed in #1334?

**Comment by xiyaowong** [comment]
Wasn't this issue fixed in #1973?

**Comment by ollien** [comment]
> Wasn't this issue fixed in #1973?

Part of it was. Though I might close this and make a new issue that's more focused, as this issue has a lot going on. The literal ask of the original works, not there are bugs sprinkled in here, as well as a feature request of showing the replacement.

**Comment by ollien** [comment]
Closing this, as #1973 addressed some of the jumpiness. #1978 and #1979 should capture the other asks in this issue. Let me know if there's something I missed!
